// appium-mac2-driver/build/lib/commands/gestures.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _appiumSupport = require("appium-support");

const commands = {};

commands.macosSetValue = async function macosSetValue(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    value,
    text,
    keyModifierFlags
  } = opts;
  return await this.wda.proxy.command(`/element/${uuid}/value`, 'POST', {
    value,
    text,
    keyModifierFlags
  });
};

commands.macosClick = async function macosClick(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/element/${uuid}/click` : '/wda/click';
  return await this.wda.proxy.command(url, 'POST', {
    x,
    y,
    keyModifierFlags
  });
};

commands.macosClickAndHold = async function macosClickAndHold(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    duration,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/element/${uuid}/clickAndHold` : '/wda/clickAndHold';
  return await this.wda.proxy.command(url, 'POST', {
    x,
    y,
    duration,
    keyModifierFlags
  });
};

commands.macosScroll = async function macosScroll(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    deltaX,
    deltaY,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/wda/element/${uuid}/scroll` : '/wda/scroll';
  return await this.wda.proxy.command(url, 'POST', {
    deltaX,
    deltaY,
    x,
    y,
    keyModifierFlags
  });
};

commands.macosRightClick = async function macosRightClick(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/wda/element/${uuid}/rightClick` : '/wda/rightClick';
  return await this.wda.proxy.command(url, 'POST', {
    x,
    y,
    keyModifierFlags
  });
};

commands.macosHover = async function macosHover(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/wda/element/${uuid}/hover` : '/wda/hover';
  return await this.wda.proxy.command(url, 'POST', {
    x,
    y,
    keyModifierFlags
  });
};

commands.macosDoubleClick = async function macosDoubleClick(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    x,
    y,
    keyModifierFlags
  } = opts;
  const url = uuid ? `/wda/element/${uuid}/doubleClick` : '/wda/doubleClick';
  return await this.wda.proxy.command(url, 'POST', {
    x,
    y,
    keyModifierFlags
  });
};

commands.macosClickAndDrag = async function macosClickAndDrag(opts = {}) {
  const sourceUuid = _appiumSupport.util.unwrapElement(opts.sourceElement);

  const destUuid = _appiumSupport.util.unwrapElement(opts.destinationElement);

  const {
    startX,
    startY,
    endX,
    endY,
    duration,
    keyModifierFlags
  } = opts;
  const url = sourceUuid && destUuid ? `/wda/element/${sourceUuid}/clickAndDrag` : '/wda/clickAndDrag';

  const dest = destUuid && _appiumSupport.util.wrapElement(destUuid);

  return await this.wda.proxy.command(url, 'POST', {
    startX,
    startY,
    endX,
    endY,
    duration,
    dest,
    keyModifierFlags
  });
};

commands.macosClickDragAndHold = async function macosClickDragAndHold(opts = {}) {
  const sourceUuid = _appiumSupport.util.unwrapElement(opts.sourceElement);

  const destUuid = _appiumSupport.util.unwrapElement(opts.destinationElement);

  const {
    startX,
    startY,
    endX,
    endY,
    duration,
    holdDuration,
    velocity,
    keyModifierFlags
  } = opts;
  const url = sourceUuid && destUuid ? `/wda/element/${sourceUuid}/clickDragAndHold` : '/wda/clickDragAndHold';

  const dest = destUuid && _appiumSupport.util.wrapElement(destUuid);

  return await this.wda.proxy.command(url, 'POST', {
    startX,
    startY,
    endX,
    endY,
    duration,
    holdDuration,
    velocity,
    dest,
    keyModifierFlags
  });
};

commands.macosKeys = async function macosKeys(opts = {}) {
  const uuid = _appiumSupport.util.unwrapElement(opts.element);

  const {
    keys
  } = opts;
  const url = uuid ? `/wda/element/${uuid}/keys` : '/wda/keys';
  return await this.wda.proxy.command(url, 'POST', {
    keys
  });
};

var _default = commands;
exports.default = _default;require('source-map-support').install();
